presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java

```diff
@@ -424,10 +424,23 @@ public static OptionalLong getTotalSizeInBytes(OptionalDouble averageColumnLengt
     /**
      * Hive calculates NDV considering null as a distinct value
      */
-    private static OptionalLong fromMetastoreDistinctValuesCount(OptionalLong distinctValuesCount, OptionalLong nullsCount)
+    private static OptionalLong fromMetastoreDistinctValuesCount(OptionalLong distinctValuesCount, OptionalLong nullsCount, OptionalLong rowCount)
     {
-        if (distinctValuesCount.isPresent() && nullsCount.isPresent() && distinctValuesCount.getAsLong() > 0 && nullsCount.getAsLong() > 0) {
-            return OptionalLong.of(distinctValuesCount.getAsLong() - 1);
+        if (distinctValuesCount.isPresent() && nullsCount.isPresent() && rowCount.isPresent()) {
+            return OptionalLong.of(fromMetastoreDistinctValuesCount(distinctValuesCount.getAsLong(), nullsCount.getAsLong(), rowCount.getAsLong()));
+        }
+        return OptionalLong.empty();
+    }
+
+    private static long fromMetastoreDistinctValuesCount(long distinctValuesCount, long nullsCount, long rowCount)
+    {
+        long nonNullsCount = rowCount - nullsCount;
+        if (nullsCount > 0) {
+            distinctValuesCount--;
+        }
+        // the metastore may store an estimate produced by the HLL, so the value stored may be higher than the total number of rows
```

Comment:
> "produced by the HLL"

or anything else, depending who wrote those stats (Hive, Spark, Impala, ...). Does Hive use HLL?

Reply:
> Does Hive use HLL?

Yes, by default. There some other, legacy sketch it can use.

Let me remove the `HLL`